### PR TITLE
[reconfig] Do not wait for batch to pick up tx in the last checkpoint

### PR DESCRIPTION
**Pull request by lxfind**
Previously we wait for the batch to pick up the newly committed transactions so that we can include them in the last checkpoint proposal in the epoch. That's wasteful and can lead to unacceptable delays when the validator is halted.
This PR makes this process proactive: we simply take care of the last step of the batch service actively and add all newly committed transactions to `extra_transactions` table.

### CODE REVIEW

File: crates/sui-core/src/authority/authority_notifier.rs
Comment (mystenmark) [inline crates/sui-core/src/authority/authority_notifier.rs:58]: Are we sure this is not a race condition? We load two values without holding a lock common to both.

I'm guessing that the answer is "the validator is halted so high_watermark cannot advance" - but is that actually true? We are still processing transactions trying to close out the epoch, right?

**Review by mystenmark — CHANGES_REQUESTED** [review]

File: crates/sui-core/src/authority/authority_notifier.rs
Comment (lxfind) [reply crates/sui-core/src/authority/authority_notifier.rs:58]: At the time this function is called (from reconfiguration), the validator is already halted, which means `high_watermark` will not move at while this function is being called. So it should be safe.
The `high_watermark` can advance latter once we start syncing the missing transactions in the last checkpoint, but that won't happen until we go pass this point.

Comment (mystenmark) [reply crates/sui-core/src/authority/authority_notifier.rs:58]: I believe you, but this seems like a very fragile invariant in that it fairly innocent changes in the future could break it. Can we instead put the loop inside this function, so that it can hold the high_watermark lock the entire time it is waiting?

Comment (lxfind) [reply crates/sui-core/src/authority/authority_notifier.rs:58]: hmm I am not sure how to do that.
There will be a async sleep in the loop, and you cannot hold a lock to the high_watermark while async wait.

Comment (mystenmark) [reply crates/sui-core/src/authority/authority_notifier.rs:58]: you can if it is a tokio mutex

Comment (lxfind) [reply crates/sui-core/src/authority/authority_notifier.rs:58]: There is an easier solution here though: I just need to hold the lock while checking the `low_watermark`, the loop can stay outside safely.

Comment (mystenmark) [reply crates/sui-core/src/authority/authority_notifier.rs:62]: Ok - approving now, I think one of those solutions would be a good idea but I don't want to hold this up any longer.

**Pull request merged**
